sql/src/main/java/io/crate/planner/operators/JoinPlanBuilder.java

```diff
@@ -0,0 +1,279 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.planner.operators;
+
+import io.crate.analyze.MultiSourceSelect;
+import io.crate.analyze.WhereClause;
+import io.crate.analyze.relations.AnalyzedRelation;
+import io.crate.analyze.relations.JoinPair;
+import io.crate.analyze.relations.QueriedRelation;
+import io.crate.analyze.relations.QuerySplitter;
+import io.crate.expression.operator.AndOperator;
+import io.crate.expression.symbol.FieldsVisitor;
+import io.crate.expression.symbol.Literal;
+import io.crate.expression.symbol.Symbol;
+import io.crate.planner.SubqueryPlanner;
+import io.crate.planner.TableStats;
+import io.crate.planner.consumer.FetchMode;
+import io.crate.planner.node.dql.join.JoinType;
+import io.crate.sql.tree.QualifiedName;
+import org.elasticsearch.common.util.set.Sets;
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+public class JoinPlanBuilder implements LogicalPlan.Builder {
```

Comment:
Please add some documentation why this is extracted (different plans NestedLoop, HashJoin, etc...)

Reply:
I think it's fairly clear what this does for now. Maybe we can later add something once there is another algorithm in use and explain how the decision is made.